File: backend/app/models/product.py
from pydantic import BaseModel, Field


class ProductSourceRecord(BaseModel):
    category: str | None = None
    source_brand_name: str | None = None
    source_brand_name_en: str | None = None
    product_name_ko: str | None = None
    product_name_en: str | None = None
    regular_price: int | None = None
    original_price: int | None = None
    sale_price: int | None = None
    discount_rate: int | None = None
    rating: float | None = None
    review_count: int | None = None
    currency: str | None = "KRW"
    shade: str | None = None
    image_url: str | None = None
    description: str | None = None
    options: list[str] | None = None
    sold_out: bool | None = None
    source: str
    source_url: str | None = None
    source_product_id: str | None = None
    updated_at: str | None = None


class ProductSearchResult(BaseModel):
    brand_ko: str | None = Field(default=None)
    brand_en: str | None = Field(default=None)
    product_name_ko: str | None = Field(default=None)
    product_name_en: str | None = Field(default=None)
    category: str | None = Field(default=None)
    price: int | None = Field(default=None)
    original_price: int | None = Field(default=None)
    sale_price: int | None = Field(default=None)
    discount_rate: int | None = Field(default=None)
    rating: float | None = Field(default=None)
    review_count: int | None = Field(default=None)
    currency: str | None = Field(default="KRW")
    shade: str | None = Field(default=None)
    image_url: str | None = Field(default=None)
    description: str | None = Field(default=None)
    options: list[str] | None = Field(default=None)
    sold_out: bool | None = Field(default=None)
    source_url: str | None = Field(default=None)
    source_product_id: str | None = Field(default=None)
    source: str
    source_label: str | None = Field(default=None)
    source_priority: int | None = Field(default=None)
    quality_score: int = Field(default=0)
    enrichment_missing_fields: list[str] = Field(default_factory=list)
    updated_at: str | None = Field(default=None)


class SearchResponse(BaseModel):
    query: str
    count: int
    results: list[ProductSearchResult]
    source_errors: list[str] = Field(default_factory=list)


class SuggestionResponse(BaseModel):
    query: str
    suggestions: list[str] = Field(default_factory=list)

File: frontend/src/types/product.ts
export type ProductOffer = {
  source: string;
  source_label?: string | null;
  source_priority?: number | null;
  source_url: string;
  source_product_id?: string | null;
  price?: number | null;
  original_price?: number | null;
  sale_price?: number | null;
  currency?: string | null;
  image_url?: string | null;
  sold_out?: boolean | null;
  updated_at?: string | null;
};

export type Product = {
  canonical_product_id?: string | null;
  brand_ko: string | null;
  brand_en: string | null;
  product_name_ko: string | null;
  product_name_en: string | null;
  product_name_display_ko?: string | null;
  product_name_display_en?: string | null;
  category?: string | null;
  price: number | null;
  original_price?: number | null;
  sale_price?: number | null;
  discount_rate?: number | null;
  rating?: number | null;
  review_count?: number | null;
  currency?: string | null;
  shade: string | null;
  image_url: string | null;
  description?: string | null;
  options?: string[] | null;
  sold_out?: boolean | null;
  source_url: string | null;
  source_product_id?: string | null;
  source: string;
  source_label?: string | null;
  source_priority?: number | null;
  quality_score?: number;
  enrichment_missing_fields?: string[];
  offers?: ProductOffer[];
  updated_at?: string | null;
};

export type SearchResponse = {
  query: string;
  count: number;
  results: Product[];
  source_errors: string[];
};

export type SuggestionResponse = {
  query: string;
  suggestions: string[];
};

export type SearchParams = {
  query: string;
  limit?: number;
};

export type EditorParsedLine = {
  raw_text: string;
  brand_query: string | null;
  brand_en: string | null;
  product_query: string | null;
  shade_query: string | null;
  shade_code: string | null;
  shade_name: string | null;
  normalized_query: string;
};

export type EditorProductCandidate = {
  product: Product;
  match_score: number;
  match_reasons?: string[];
};

export type EditorBatchItem = {
  raw_text: string;
  parsed: EditorParsedLine;
  status: "확인됨" | "후보 있음" | "수동 확인 필요";
  candidates: EditorProductCandidate[];
};

export type EditorBatchResponse = {
  count: number;
  items: EditorBatchItem[];
};

export type EditorBatchProgress = {
  completed: number;
  total: number;
  response: EditorBatchResponse;
};

